examples/lottery/src/lib.rs

```diff
@@ -0,0 +1,346 @@
+            
+            if sum != PERCENTAGE_TOTAL {
+                return sc_error!("Prize distribution must add up to exactly 100(%)!");
+            }
+        }
+    
+        self.set_ticket_price(&lottery_name, ticket_price);
+        self.set_tickets_left(&lottery_name, tt);
+        self.set_deadline(&lottery_name, d);
+        self.set_max_entries_per_user(&lottery_name, max);
+        self.set_prize_distribution(&lottery_name, &pd);
+        self.set_whitelist(&lottery_name, &wl);
+    
+        Ok(())
+    }
+
+    #[endpoint]
+    #[payable]
+    fn buy_ticket(&self, lottery_name: Vec<u8>, #[payment] _payment: BigUint) -> SCResult<()> {
+        match self.status(lottery_name.clone()) {
+            Status::Inactive => {
+                sc_error!("Lottery is currently inactive.")
+            },
+            Status::Running => {
+                let whitelist = self.get_whitelist(&lottery_name);
+                let caller = self.get_caller();
+
+                if !whitelist.is_empty() && !whitelist.contains(&caller) {
+                    return sc_error!("You are not allowed to participate in this lottery!");
+                }
+
+                let ticket_price = self.get_ticket_price(&lottery_name);
+
+                if _payment != ticket_price {
+                    return sc_error!("Wrong ticket fee!");
+                }
+
+                let mut entries = self.get_number_of_entries_for_user(&lottery_name, &caller);
+                let max_entries = self.get_max_entries_per_user(&lottery_name);
+
+                if *entries == max_entries {
+                    return sc_error!("Ticket limit exceeded for this lottery!");
+                }
+
+                let mut ticket_number = self.get_mut_current_ticket_number(&lottery_name);
+                let mut tickets_left = self.get_mut_tickets_left(&lottery_name);
+                let mut prize_pool = self.get_mut_prize_pool(&lottery_name);
+
+                self.set_ticket_holder(&lottery_name, *ticket_number, &caller);
+                *entries += 1;
+                *ticket_number += 1;
+                *tickets_left -= 1;
+                *prize_pool += ticket_price;
+
+                Ok(())
+            },
+            Status::Ended => {
+                sc_error!("Lottery entry period has ended! Awaiting winner announcement.")
+            }
+        }
+    }
+
+    #[endpoint]
+    fn determine_winner(&self, lottery_name: Vec<u8>) -> SCResult<()> {
+        match self.status(lottery_name.clone()) {
+            Status::Inactive => { 
+                sc_error!("Lottery is inactive!")
+            },
+            Status::Running => {
+                sc_error!("Lottery is still running!")
+            },
+            Status::Ended => {
+                let total_tickets = self.get_mut_current_ticket_number(&lottery_name);
+
+                if *total_tickets > 0 {
+                    let mut seed = self.get_block_nonce() + self.get_block_timestamp();
```

Comment:
this is not a good random - this can be played and tricked by the leader. Please use the randomness given by the API.
There are so called forward attacks - where one leader will try to execute a set of transactions and he can try several times until time is out for processing this block - until he finds a combination where he wins.